zebra/zebra_etcd.c

```diff
@@ -0,0 +1,639 @@
+
+  iter = &zetcd_glob_p->t_conn_up_state.iter;
+  /*Iterate over route nodes and trigger update for each entry.*/
+  while ((rnode = zetcd_rnodes_iter_next(iter))){
+    dest = rib_dest_from_rnode(rnode);
+    if (dest){
+      zetcd_trigger_update(rnode,NULL);
+    }
+  }
+  return 0;
+}
+/**
+ * zetcd_connection_up
+ *
+ * After establishing connection initialize the route node iterator and add 
+ * task zetcd_conn_up_thread_cb in timer thread to process the route nodes.
+ *
+ * */
+static void zetcd_connection_up(const char *reason)
+{
+  assert(&zetcd_glob_p->channel);
+  assert(!zetcd_glob_p->t_conn_up);
+  zetcd_rnodes_iter_init(&zetcd_glob_p->t_conn_up_state.iter);
+  thread_add_timer_msec(zetcd_glob_p->master, zetcd_conn_up_thread_cb, NULL, 0,
+            &zetcd_glob_p->t_conn_up);
+}
+
+/**
+ * zetcd_connect_cb
+ * 
+ * create a grpc channel and put and delete a sample to check connectivity. If
+ * connectivity failed try to reconnect. If unable to connect after
+ * ZETCD_CONNECT_RETRY_ATTEMPT then abort.
+ * 
+ * @param[in] txn_put_req a pointer to etcd put transaction request object.
+ * @param[in] txn_del_req a pointer to etcd delete transaction request object.
+ * 
+ * */
+
+static int zetcd_connect_cb(struct thread *t)
+{
+  int ret;
+
+  /*Check if already connected.*/
+  if (zetcd_glob_p->is_connected)
+    return 0;
+
+  /*Check if max attempt exceeded.*/
+  if (zetcd_glob_p->conn_retries > ZETCD_CONNECT_RETRY_ATTEMPT)
+  {
+    zlog_warn("Max attempt to connect to etcd server exceeded.");
+    return -1;
+  }
+
+  /* Create sample put object.*/
+  create_put_txn("sample","test",10,zetcd_glob_p->txn_put_req);
+
+  /*Create grpc channel using wrapper function with zetcd_server URI.*/
+	create_grpc_channel_(zetcd_glob_p->zetcd_server);
+
+  /*Carry out the put transaction to put "sample" key into etcd server.*/
+  ret = etcd_txn(zetcd_glob_p->txn_put_req);
+  if(ret)
+  {
+    zlog_err("error Unable to push into etcd");
+    zetcd_glob_p->conn_retries++;
+    /*If failed try another attempt.*/
+    zetcd_start_connect_timer("connection failed");
+    return -1;
+  }
+  /*delete opration.*/
+  create_del_txn("sample", false, zetcd_glob_p->txn_put_req);
+  ret = etcd_txn(zetcd_glob_p->txn_del_req);
+  if(ret)
+  {
+    zlog_err("error Unable to push into etcd");
+    zetcd_glob_p->conn_retries++;
+    zetcd_start_connect_timer("connection failed");
+    return -1;
+  }
+	zetcd_glob_p->is_connected = true;
+	zetcd_connection_up("connection successful");
+	return 0;
+}
+
+/**
+ * zetcd_start_connect_timer
+ *
+ * Schedule a zetcd_connect_cb task after some assertion.
+ *
+ * @param[in] t_connect pointer to timer thread object to connect to etcd-server.
+ * @param[in] is_connected flag to check the connectivity.
+ * */
+static void zetcd_start_connect_timer(const char *reason)
+{
+  zlog_debug("etcd connection thread started as :%s",reason);
+  assert(!zetcd_glob_p->t_connect);
+  assert(!zetcd_glob_p->is_connected);
```

Comment:
The assertion was to avoid scheduling of the task to connect to `etcd server`, if it is already another same task scheduled in queue.
Later I realized if it asserts then program aborts, which is not expected. So, yes `if (zetcd_glob_p->t_connect) { return scheduled}` will suffice.